pyogrio/_io.pyx

```diff
@@ -2178,3 +2180,182 @@ def ogr_write(
         exc = exc_check()
         if exc:
             raise DataSourceError(f"Failed to write features to dataset {path}; {exc}")
+
+
+def ogr_write_arrow(
+    str path,
+    str layer,
+    str driver,
+    arrow_obj,
+    str crs,
+    str geometry_type,
+    str geometry_name,
+    str encoding,
+    object dataset_kwargs,
+    object layer_kwargs,
+    bint append=False,
+    dataset_metadata=None,
+    layer_metadata=None,
+):
+    IF CTE_GDAL_VERSION < (3, 8, 0):
+        raise RuntimeError("Need GDAL>=3.8 for Arrow write support")
+
+    cdef OGRDataSourceH ogr_dataset = NULL
+    cdef OGRLayerH ogr_layer = NULL
+
+    if not encoding:
+        encoding = locale.getpreferredencoding()
+
+    layer_created = create_ogr_dataset_layer(
+        path, layer, driver, crs, geometry_type, encoding,
+        dataset_kwargs, layer_kwargs, append,
+        dataset_metadata, layer_metadata,
+        &ogr_dataset, &ogr_layer,
+    )
+    if geometry_name:
```

Comment:
`dict_to_options` already encodes it as UTF-8, always. But does that mean `dict_to_options` should be updated to not always do that, and to actually pass it an encoding?

Reply:
Sorry for the confusion, what I meant was this block for the dataset-level encoding, after the `layer_created = create_ogr_dataset_layer(...)` block:

```Python
    if driver == 'ESRI Shapefile':
        # force encoding for remaining operations to be in UTF-8 (even if user
        # provides an encoding) because GDAL will automatically convert those to
        # the target encoding because ENCODING is set as a layer creation option
        encoding = "UTF-8"

    else:
        # Now the dataset and layer have been created, we can properly determine the
        # encoding. It is derived from the user, from the dataset capabilities / type,
        # or from the system locale
        encoding = encoding or detect_encoding(ogr_dataset, ogr_layer)
```

`dict_to_options` should always encode keys and values to `UTF-8` before sending to GDAL, since those are used by GDAL rather than written out based on the target encoding of the dataset.

Reply:
`encoding` isn't used anymore after the dataset/layer has been created. I am not entirely sure how GDAL handles that for writing the data through the Arrow API for a dataset/layer that isn't in UTF-8, though

Reply:
Ah, good point!  We ran into the same issue reading Arrow data as well, and pyarrow breaks badly (as we have it now) if the data read from the data source is not UTF-8 (e.g., if we forced it to be written that way for a non-shapefile driver)